melonproject/protocol
# Refactor: Update deploy scripts and tests to run on both local and testnets

Addresses issues #843  and #844 .

This PR essentially does two things:

1. Refactor deployment scripts so that they can deploy to either local blockchain or a testnet

2. Refactor tests so that they can run - when possible - on either local blockchain or a testnet

It also cleans up tests a bit for consistency and readability

Some general notes for other devs working on the project:


DEPLOYMENT NOTES
- Add kovan commands to package.json (`deploy-kovan` and `test-kovan`). `test-kovan` will run all the tests in any folder without `local/` in its path
- Add `deploy_in_kovan` config (and update `deploy_in_testing` with some changes).
- Add to kovan_keys so we have a total of 5 accounts for tests


TEST REFACTORING NOTES
- Refactor all contract methods to use `call` and `send` helpers rather than calling directly in tests (this gives us more control of how we call/send txs, and @travs noted that this will be helpful when we transition to web3 2.x)
- Refactor assertions to use `bigNumberEq` where possibly
- Add a `bigNumberCloseTo` jest expect method, since sometimes our estimates vs expected values are like `1999999999999` and `2000000000000`. This method takes a `margin` as a param, so we can define what an acceptable offset is.
- Substitute a `delay` helper (instead of `increaseTime`) to force short realtime pauses, so that we can affect time between blocks without forcing EVM manipulation
- Refactor all fund-related helpers into `fund.js`. Most new tests should use `setupFundWithParams` from now on.
- Tests in the `integration/local/legacy` folder won't be updated with new syntax, as they don't have clear value-adds to our other integration tests.
- Add a `toThrowFlexible` jest expect method, to allow either a reason string match if a local blockchain, or a general error if using Infura.
